test/Microsoft.Framework.DesignTimeHost.Tests/PluginHandlerFacts.cs

```diff
@@ -0,0 +1,491 @@
+                    { "TypeName", typeof(TestPlugin).FullName },
+                },
+                MessageName = "RegisterPlugin",
+                PluginId = RandomGuidId
+            };
+            var unregisterPluginMessage = new PluginMessage
+            {
+                MessageName = "UnregisterPlugin",
+                PluginId = RandomGuidId
+            };
+            var expectedErrorMessage = 
+                $"No plugin with id '{RandomGuidId}' has been registered. Cannot unregister plugin.";
+
+            pluginHandler.ProcessMessage(registerPluginMessage, assemblyLoadContext);
+            pluginHandler.ProcessMessage(unregisterPluginMessage, assemblyLoadContext);
+
+            var error = Assert.Throws<InvalidOperationException>(
+                () => pluginHandler.ProcessMessage(unregisterPluginMessage, assemblyLoadContext));
+            Assert.Equal(expectedErrorMessage, error.Message, StringComparer.Ordinal);
+            error = Assert.Throws<InvalidOperationException>(
+                () => pluginHandler.ProcessMessage(unregisterPluginMessage, assemblyLoadContext));
+            Assert.Equal(expectedErrorMessage, error.Message, StringComparer.Ordinal);
+        }
+
+        [Fact]
+        public void ProcessMessage_PluginMessage_ProcessesMessages()
+        {
+            var assemblyLoadContext = CreateTestAssemblyLoadContext<MessageTestPlugin>();
+            var serviceProvider = new TestServiceProvider();
+            PluginMessageBroker.PluginMessageWrapperData messageBrokerData = null;
+            var pluginHandler = new PluginHandler(
+                serviceProvider, (data) => messageBrokerData = (PluginMessageBroker.PluginMessageWrapperData)data);
+            var registerPluginMessage = new PluginMessage
+            {
+                Data = new JObject
+                {
+                    { "AssemblyName", "CustomAssembly" },
+                    { "TypeName", typeof(MessageTestPlugin).FullName },
+                },
+                MessageName = "RegisterPlugin",
+                PluginId = RandomGuidId
+            };
+            var pluginMessage = new PluginMessage
+            {
+                Data = new JObject
+                {
+                    { "Data", "Hello Plugin" },
+                },
+                MessageName = "PluginMessage",
+                PluginId = RandomGuidId
+            };
+
+            pluginHandler.ProcessMessage(registerPluginMessage, assemblyLoadContext);
+            pluginHandler.ProcessMessage(pluginMessage, assemblyLoadContext);
+
+            Assert.NotNull(messageBrokerData);
+            Assert.Equal(RandomGuidId, messageBrokerData.PluginId);
+            var actualMessage = (string)messageBrokerData.Data;
+            Assert.Equal("Hello Plugin!", actualMessage, StringComparer.Ordinal);
+        }
+
+        [Fact]
+        public void ProcessMessage_PluginMessage_ThrowsWhenPluginNotRegistered()
```

Comment:
interesting behaviour.  many messaging systems simply ignore messages when destination hasn't been registered.  a rare few store (some) messages in case the destination is registered (soonish).  why is `PluginHandler` so strict when sending messages (but ignore invalid registration attempts)?